### ISSUE TITLE
Add immutable dequeue

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#3115]
See e.q. https://javadoc.io/static/org.scalaz/scalaz_2.9.3/7.1.17/scalaz/Dequeue.html

**Comment by magnus-madsen** [comment]
Yes, I would suggest to write the first 100 lines and then make a PR so we can discuss the architecture.

You might also want to find some papers etc on how to best implement it.

Note that for all public functions we expect 3-5 test cases.

**Comment by magnus-madsen** [comment]
Sure. I'm no expert either.

**Comment by magnus-madsen** [comment]
Following okasaki is also good

**Comment by magnus-madsen** [comment]
We do have `Lazy[t]`, but are you saying Deque cannot be done without?

In any case, https://github.com/flix/flix/issues/4533 is an experiment, and probably not something we should build on top of...

We do have `DelayList` though.

**Comment by magnus-madsen** [comment]
I think that experimentation is important, so I encourage you to go forward :)

As for getting something merged into Flix, I think the *first* task would simply be to agree on the API (i.e. list of public functions). Once that is in place, we can experiment with different implementations, right? :)

**Comment by magnus-madsen** [comment]
Makes sense?

**Comment by magnus-madsen** [comment]
I don't think we should start with a type class; rather I would identify the central operations, implement those, and then we can expand.
